src/utilities/columns.ts

```diff
@@ -0,0 +1,143 @@
+import { RunGraphData, RunGraphNodes } from '@/models/RunGraph'
+
+/**
+ * A `Map` object that maps each node ID to its column.
+ */
+export type NodeColumns = Map<string, number>
+
+/**
+ * Calculates the column for each node in the run graph using a faster or slower algorithm.
+ *
+ * This function first tries to calculate the columns using the faster `getColumnsFaster` function. If an error occurs during the calculation, it falls back to the slower `getColumnsSlower` function. The function takes in a `RunGraphData` object containing the root node IDs and nodes in the run graph.
+ *
+ * @param runGraphData - A `RunGraphData` object containing the root node IDs and nodes in the run graph.
+ * @returns A `Map` object that maps each node ID to its column.
+ */
+export function getColumns({ root_node_ids, nodes }: RunGraphData): NodeColumns {
+  try {
+    return getColumnsFaster(root_node_ids, nodes)
+  } catch (error) {
+    console.error(error)
+    return getColumnsSlower(root_node_ids, nodes)
+  }
+}
+
+/**
+ * Calculates the column for each node in the run graph using a faster algorithm.
+ *
+ * This function sets up the initial columns for the root nodes and then calls `getChildrenColumns` to process all the children.
+ *
+ * Performance: 200k operations per seconds processing 23 connected nodes
+ * Performance: 54-58 operations per second processing 2,000 highly connected nodes.
+ *
+ * @param nodeIds - The IDs of the root nodes.
+ * @param nodes - The nodes in the run graph.
+ * @returns A `Map` object that maps each node ID to its column.
+ */
+function getColumnsFaster(nodeIds: string[], nodes: RunGraphNodes): NodeColumns {
+  const columns: NodeColumns = new Map()
+  const childrenNodeIds: string[] = []
+
+  for (const id of nodeIds) {
+    columns.set(id, 0)
+    const node = nodes.get(id)!
+
+    for (const { id: childId } of node.children) {
+      childrenNodeIds.push(childId)
+    }
+  }
+
+  return getChildrenColumns(childrenNodeIds, nodes, columns)
+}
+
+/**
+ * Calculates the column for each child node of the given nodes.
+ *
+ * This function processes the children of each node in the `nodeIds` array and calculates their columns based on the columns of their parents.
+ *
+ * @param nodeIds - The IDs of the nodes whose children should be processed.
+ * @param nodes - The nodes in the run graph.
+ * @param columns - A `Map` object that maps each node ID to its column.
+ * @returns A `Map` object that maps each node ID to its column.
+ * @throws An error if a node ID in `nodeIds` is not found in `nodes`.
+ */
+function getChildrenColumns(nodeIds: string[], nodes: RunGraphNodes, columns: NodeColumns): NodeColumns {
```

Comment:
Could this be an async function to avoid the stack overflow?

Reply:
You said this earlier and its just clicking with me what you mean. Let me test some things

Reply:
Discussed this via slack. But making this async cuts its performance over half. Still faster than the "getColumnsSlower" method though. 

Might optimize this further but for now I want to move on to other pieces.